### ISSUE TITLE
OS X .whl install results in ssl verification failures

### ISSUE THREAD

**Issue by ldanielburr** [pyca/cryptography#2554]
OS X Yosemite 10.10.5

Recently updated to cryptography 1.1.2, and was hugely pleased to see that a .whl was installed (cryptography-1.1.2-pp27-none-macosx_10_10_x86_64.whl), rather that having to do the macports dance that I've done in the past.

Then, bad things started happening whenever my code made HTTPS requests:
OpenSSL.SSL.Error: [('SSL routines', 'ssl3_get_server_certificate', 'certificate verify failed')]>]

Long story short, I reinstalled cryptography, forcing it to build against my macports OpenSSL:
env LDFLAGS="-L/opt/local/lib" CFLAGS="-I/opt/local/include" pip install --no-cache-dir cryptography --no-binary :all:

After this, everything works again, but I've no idea why.

**Comment by reaperhulk** [comment]
This is almost certainly related to the statically linked OpenSSL we have trying (and failing) to open a trust store that doesn't exist on your machine. We are going to need to figure out what to do about this. Hopefully the answer is something like "add certifi to our dependencies and have OpenSSL use that by default". cc @Lukasa

**Comment by reaperhulk** [comment]
Complicating this a bit is whether we want to override system trust store on linux (where, since the user is compiling anyway, the path is likely to be correct). If we don't then we need a conditional dependency on certifi, but only for the platforms where we ship static linked OpenSSL in binary wheels (Windows and OS X).

**Comment by ldanielburr** [comment]
@reaperhulk - Thanks for explaining, it makes sense.  Where is the statically linked OpenSSL looking for the trust store?  I can at least verify that placing a trust store in the right location causes the .whl to work as expected.

**Comment by nbastin** [comment]
If you can find out where it's looking for the trust store, you can maybe at least do something horrible like:

    security find-certificate -a -p /System/Library/Keychains/SystemCACertificates.keychain > systemca.pem

And put that output file where OpenSSL wants to find it.  (This doesn't cover other keychains, of course, although you could trivially expand to all of them).

**Comment by Lukasa** [comment]
OpenSSL can be told where to find the trust store as [a compile-time option](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/openssl.rb#L45), which means we can get this 'right' using the wheel builders. 

In this case, I think the most sensible place to put it is where Homebrew puts it, and hope that Homebrew OpenSSL has been installed to correctly populate that directory, but that may be a bit aggressive against MacPorts users.

**Comment by reaperhulk** [comment]
That's effectively what we're doing right now (which is why @ldanielburr is seeing the problem, as he's a MacPorts user). I'm reluctant to have an implicit dependency on homebrew openssl when we static link to avoid that very thing.

**Comment by ldanielburr** [comment]
@Lukasa - Macports supports the same [compile-time option](https://trac.macports.org/browser/trunk/dports/devel/openssl/Portfile#L51), for what it is worth.

**Comment by nbastin** [comment]
@reaperhulk : You can also tell it where to find the trust store as a run-time option, which we could let the user do as a fallback.

**Comment by Lukasa** [comment]
@reaperhulk In that case, it gets tricky fast. One option would be to depend on certifi and then point at the install directory for certifi, but that will only work for system packages, not virtualenvs.

Otherwise, you *could* go the requests route and vendor certifi's PEM bundle directly. This will give you *yet another* situation in which you may want to version bump at short-notice, but will guarantee that you've always got the bundle available at a sensible location relative to openssl. The only issue *there* might be that openssl's openssdir may be resolved into an absolute path at compile time, in which case once again depending on certifi doesn't work for virtualenvs.

**Comment by alex** [comment]
Ugh, I don't really want to be in the business of being responsible for a
trust store.

On Wed, Dec 23, 2015 at 2:28 PM, Cory Benfield <notifications@github.com>
wrote:

> @reaperhulk <https://github.com/reaperhulk> In that case, it gets tricky
> fast. One option would be to depend on certifi and then point at the
> install directory for certifi, but that will only work for system packages,
> not virtualenvs.
>
> Otherwise, you *could* go the requests route and vendor certifi's PEM
> bundle directly. This will give you *yet another* situation in which you
> may want to version bump at short-notice, but will guarantee that you've
> always got the bundle available at a sensible location relative to openssl.
> The only issue *there* might be that openssl's openssdir may be resolved
> into an absolute path at compile time, in which case once again depending
> on certifi doesn't work for virtualenvs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/issues/2554#issuecomment-166975932>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by Lukasa** [comment]
@alex For good reason.

**Comment by ldanielburr** [comment]
If the wheel installation works as expected for OS X users who install openssl via homebrew, and it is only macports (or fink) users who have this problem, perhaps it could just be addressed in the documentation, so that people like myself could be aware, and continue to the the non-binary installation.  That would have solved my problem right away, particularly the pip syntax needed to explicitly force the compile-and-link-against-my-openssl installation.

**Comment by cjw296** [comment]
Curious where this issue has got to?

I'm in the unfortunately position of using Python 2.7 on OS, installed from Python.org, not using homebrew or macports and don't really want to, where I need a private CA cert to access a bunch of our internal SSL-fronted web-apps.

I hit this first up with pip where, once I have ndg-httpsclient installed in a virtualenv, I get "bad handshake: Error([('SSL routines', 'ssl3_get_server_certificate', 'certificate verify failed')],)" when trying to install anything from our internal devpi respository (which is served over SSL used a cert signed by our internal CA).

How can I get ndg-httpsclient/PyOpenSSL which I suspect both end up hitting this cryptography library, to use the OS X internal certificate stores? (rather than needing a .pem somewhere...)

**Comment by Lukasa** [comment]
@cjw296 The short answer is that you cannot do it easily. OpenSSL knows nothing about the OS X internal cert stores, and even if it did, simply passing those certs to OpenSSL can lead to unexpected results. The [certitude](https://github.com/python-hyper/certitude) library can in principle be used to validate those cert chains, but requests/urllib3 would need substantial changes to be able to use it.

**Comment by cjw296** [comment]
@Lukasa - https://www.mercurial-scm.org/wiki/CACertificates#Mac_OS_X_10.6_and_higher suggests Apple got their OpenSSL to talk to their internal stores. Would I be right in guessing that's the OpenSSL they're no longer supporting?

**Comment by Lukasa** [comment]
@cjw296 It is indeed. It's also very much not something you actually want to use: it breaks in other circumstances, per https://hynek.me/articles/apple-openssl-verification-surprises/

**Comment by cjw296** [comment]
Right, so what do modern OSX users on OpenSSL do? Having multiple trust stores seems like a bad idea all round...

**Comment by Lukasa** [comment]
@cjw296 It is. However, that's usually what they do. The Homebrew installed OpenSSL exports the roots at install time: not a great solution, but ok. In the short term I recommend you do something similar.

**Comment by Lukasa** [comment]
For what it's worth, @cjw296, I'm actively looking in to how to improve this situation. Certitude is one possible route, but I'm also entertaining the possibility of defining a common TLS API for Python and then making wrappers for various TLS implementations. That would also vastly help for Python on OS X.

**Comment by trenton42** [comment]
I noticed that this issue is still open, and I ran into this again today. Strangely, installing the `1.8` wheel works fine, but `1.8.1` does fails.

**Comment by alex** [comment]
That's probably because `1.8.1` switched to linking against OpenSSL 1.1.0; I bet a `brew install openssl@1.1` will fix it :-/

**Comment by trenton42** [comment]
Thanks @alex, that indeed fixed it! I wasn't paying enough attention to notice that homebrew had both an `openssl` and an `openssl@1.1`

**Comment by reaperhulk** [comment]
I think we should turn this into a documentation task for pyOpenSSL. https://github.com/pyca/pyopenssl/issues/642
